webhook/resources/replica/validator.go

```diff
@@ -70,3 +71,51 @@ func (r *replicaValidator) Update(request *admission.Request, oldObj runtime.Obj
 
 	return nil
 }
+
+func (r *replicaValidator) Delete(request *admission.Request, oldObj runtime.Object) error {
+	replica := oldObj.(*longhorn.Replica)
+
+	if err := r.validateReplicaDeletion(replica); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	return nil
+}
+
+func (r *replicaValidator) validateReplicaDeletion(replica *longhorn.Replica) error {
+	if replica.Spec.VolumeName == "" {
+		return nil
+	}
+
+	volume, err := r.ds.GetVolumeRO(replica.Spec.VolumeName)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to get volume %v before deleting replica", replica.Spec.VolumeName)
+	}
+
+	if volume.DeletionTimestamp != nil {
+		return nil
+	}
+
+	replicas, err := r.ds.ListVolumeReplicasRO(volume.Name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to list replicas for volume %v before deleting replica", volume.Name)
+	}
+
+	healthyReplicas := map[string]struct{}{}
+	for _, r := range replicas {
+		if r.Spec.HealthyAt != "" {
+			healthyReplicas[r.Name] = struct{}{}
+		}
+	}
+	if len(healthyReplicas) == 1 {
+		if _, ok := healthyReplicas[replica.Name]; !ok {
+			return nil
+		}
+		return fmt.Errorf("cannot delete replica %v because volume %v only has one available replica", replica.Name, volume.Name)
+	}
+
+	return nil
```

Comment:
`To prevent accidentally from deleting the last replica regardless of it is healthy or faulty`

It seems we did not handle if the last replica is failed here?

Reply:
If a volume only has one available replica whose `healthyAt != ""` but `failedAt != ""` , then we cannot allow the deletion. The replica might be the last hope of the volume.